### PR TITLE
Remember Popped-up Chat Size

### LINKED ISSUE #1788: Remember Popped-up Chat Size
**What should be added?**
Add a feature that would save the size of a popped up chat. 
**Why should it be added?**
Right now whenever I popup a chat, its size is always the same, and it's very small so I have to resize it every time. Would be nice if there was a feature that would either keep the size I set it to last time or maybe a setting that would let me set a specific size (for eg. X by Y pixels).

### PR DESCRIPTION
**Pull request by maliByatzes**
Fixes #1788
Adedd two new values in the `Settings.hpp` file to keep track of the last popup window's width and height. On closing a popup window, while the main window is still active, both `lastPopupWidth` and `lastPopupHeight` are updated with the last window's dimensions. When a new popup window is created it uses those same dimensions to be same width and height as the last window.

### CODE REVIEW

File: src/singletons/Settings.hpp [hunk 1]
```diff
@@ -212,6 +212,8 @@ class Settings
     BoolSetting useCustomFfzVipBadges = {
         "/appearance/badges/useCustomFfzVipBadges", true};
     BoolSetting showBadgesSevenTV = {"/appearance/badges/seventv", true};
+    IntSetting lastPopupWidth = {"/appearance/popup/lastWidth", 300};
+    IntSetting lastPopupHeight = {"/appearance/popup/lastHeight", 500};
```
[the remark this answers is not in the document]
Comment (pajlada) [reply src/singletons/Settings.hpp:216]: If you're up for it, this would be a nice addition

**Comment by pajlada** [comment]
> (discussion): Should we only save the dimensions?

imo yeah, at least as part of this change. keeping it simple

**Comment by maliByatzes** [comment]
> Thank you @maliByatzes!
> 
> Two open things:
> 
> 1. As Nerixyz mentioned in [Remember Popped-up Chat Size #5635 (comment)](https://github.com/Chatterino/chatterino2/pull/5635#discussion_r1792093298) - If you're up for it, you can change the setting to be a `ChatterinoSetting<QSize>`, meaning you'd need to implement a custom serializer/deserializer as in https://github.com/Chatterino/chatterino2/blob/afa8067a20e40df6260d1b339624b638221da783/src/controllers/highlights/HighlightBadge.hpp#L74-L136
>     - if this is something you'd rather not do, let me know and I'll help you implement it in this PR
> 2. Look at the scaling thing Nerixyz mentioned in [Remember Popped-up Chat Size #5635 (comment)](https://github.com/Chatterino/chatterino2/pull/5635#discussion_r1792077570) and ensure the size is retained as expected

Thank you for your feedback, I'll look into these suggested changes, I will scream for help if I can't figure it out.

**Comment by maliByatzes** [comment]
@pajlada  I wrote the below Serialize and Deserialize for QSize, I'm wondering if its corrects?

```
#pragma once

#include <pajlada/serialize.hpp>
#include <QSize>

namespace pajlada {

template <>
struct Serialize<QSize> {
    static rapidjson::Value get(const QSize &value,
                                rapidjson::Document::AllocatorType &a)
    {
        rapidjson::Value ret(rapidjson::kObjectType);

        chatterino::rj::set(ret, "width", value.width(), a);
        chatterino::rj::set(ret, "height", value.height(), a);

        return ret;
    }
};

template <>
struct Deserialize<QSize> {
    static QSize get(const rapidjson::Value &value, bool *error)
    {
        if (!value.IsObject())
        {
            PAJLADA_REPORT_ERROR(error);
            return QSize{};
        }

        int _width;
        int _height;

        chatterino::rj::getSafe(value, "width", _width);
        chatterino::rj::getSafe(value, "height", _height);

        return QSize(_width, _height);
    }
};

}  // namespace pajlada
```

**Comment by pajlada** [comment]
> @pajlada I wrote the below Serialize and Deserialize for QSize, I'm wondering if it's corrects?

That looks correct at a glance yeah!

**Comment by maliByatzes** [comment]
@pajlada Im having a problem with the changes I made above, the `lastPopup` always return 300 and 500 it is not updated. So im wondering if the `.getValue()` method is a copy not a reference.

**Comment by pajlada** [comment]
> Im having a problem with the changes I made above, the `lastPopup` always return 300 and 500 it is not updated. So im wondering if the `.getValue()` method is a copy not a reference or it is just not updated at all.

Yeah you're right - `getValue` returns a copy, not a reference.
So at [`5755dc0` (#5635)](https://github.com/Chatterino/chatterino2/pull/5635/commits/5755dc02b0e18c0ea460c3fdc0812aa0a6007845#diff-fb5d21ad3648166c00d0dbfd39932dc9ed39eff87f35a1aa90312c0bc78d0e98R155-R157) you need to create a QSize then call `setValue` on the setting with the QSize you just created

**Comment by maliByatzes** [comment]
@pajlada  A call to `setValue` is returning the below error:
```
/home/malib/software/chatterino2/lib/serialize/include/pajlada/serialize/serialize.hpp:36:17: error: no matching function for call to ‘rapidjson::GenericValue<rapidjson:
:UTF8<> >::GenericValue(const QSize&)’                                                                                                                                   
   36 |         RJValue ret(value);                                                                                                                                      
      |                 ^~~           
```
with the below snippet using `setValue`
```
QRect rect = this->getBounds();                                                                                                                                  
QSize newSize(rect.width(), rect.height());                                                                                                                      
getSettings()->lastPopupSetting.setValue(newSize); 
```

**Comment by pajlada** [comment]
@maliByatzes Any place that sets lastPopupSetting will need to include the new file you made (`RapidJsonSerializeQSize`) so the serialize library can find the implementation

It *could* be included in `Settings.hpp` like the QString variant, but because this is so rarely used, I think I'd prefer if it's only included in `Window.cpp`

**Comment by pajlada** [comment]
I'd prefer to not do position tracking for now, since I'm not sure it's something we want

**Review by pajlada — APPROVED** [review]
Thank you! As a first-time contributor, you are now eligible to be listed in the About page in Chatterino.

If you want this, you can open a new PR where you modify the `resources/contributors.txt` file and add yourself to the list. Read the comments at the top for instructions.

Let me know if you have any questions about the follow-up commits I made to your code

**Pull request merged**